### PR TITLE
INFRA-1590 - add several useful functions/fixes to boto_asg, boto_iam, boto_lc

### PR DESCRIPTION
**Pull request by tkwilliams**
New functions:
- modules/boto_asg.py:  get_all_launch_configurations(), list_launch_configurations(), describe_launch_configuration()
- modules/boto_iam.py:  get_all_instance_profiles(), list_instance_profiles()

Updated functions:
- states/boto_asg.py:  add 'subnet_names' param to present(), add example to clarify non-obvious launch_config param syntax, fix bug I previously introduced with the 'image_name' param, fix potential 'is not None' bug where original case would also erroneously match if sg_index == 0
- states/boto_lc.py:  update docstring for present() to indent sub-params which are NOT top-level and will in fact fail the call if passed in the way they were previously detailed.

### CODE REVIEW

File: salt/modules/boto_asg.py [hunk 1]
```diff
@@ -499,6 +499,59 @@ def launch_configuration_exists(name, region=None, key=None, keyid=None,
...
+
+    CLI example::
+
+        salt myminion boto_asg.get_all_launch_configurations
+    '''
+    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)
+    try:
+        return conn.get_all_launch_configurations()
+    except boto.exception.BotoServerError as e:
+        log.debug(e)
```
Comment (cachedout) [inline salt/modules/boto_asg.py]: This will silently fail for most people. Should we raise this to `log.error`?

File: salt/modules/boto_asg.py [hunk 2]
```diff
@@ -499,6 +499,59 @@ def launch_configuration_exists(name, region=None, key=None, keyid=None,
...
+        salt myminion boto_asg.describe_launch_configuration mylc
+    '''
+    conn = _get_conn(region=region, key=key, keyid=keyid, profile=profile)
+    try:
+        lc = conn.get_all_launch_configurations(names=[name])
+        if lc:
+            return lc[0]
+        else:
+            msg = 'The launch configuration does not exist in region {0}'.format(region)
+            log.debug(msg)
```
Comment (cachedout) [inline salt/modules/boto_asg.py]: Same question here about log level.

**Comment by cachedout** [comment]
Hi @tkwilliams . Just some questions about log levels here. Other than those, this looks ready to go. Thanks so much!

**Comment by tkwilliams** [comment]
@cachedout  Yah, you're absolutely right :)  Fixed.  I also updated the error-case return value of get_all_launch_configurations() since it's called by list_...() and the latter always expects an iterable.

**Comment by tkwilliams** [comment]
Actually, with recent versions, pretty much EVERYTHING fails silently for most users, since the default log-level is now apparently "error" instead of "info", even for salt-call runs....  :)

**Comment by tkwilliams** [comment]
@cachedout  - I /think/ this is ready to rock now.  I've updated the bits you asked about, and fixed up the merge conflict that had arisen since I created the branch, so it should be good to go.

Thanks!
t.

**Pull request merged**